### ISSUE TITLE
wxmaxima crashes when pasting while something is selected

### ISSUE THREAD

**Issue by MSoegtropIMC** [wxMaxima-developers/wxmaxima#1309]
wxmaxima crashes in the following situation:

- select some cells (e.g. a text cell) and press Ctrl+C / Cmd+C
- select some other cell and press Ctrl+V / Cmd+V

as far as I can tell this happens always regardless of the cells and history. E.g. I created a new document, inserted two text cells with Ctrl+1 / Cmd+1, selected and copied the first and pasted it over the second.

```
wxMaxima is a graphical front end for Maxima, which does the mathematics in the background.
If you encounter a mathematical problem, it is probably a Maxima bug und should be submitted there.
The Maxima bug database is available at
    https://sourceforge.net/p/maxima/bugs
Submit bug reports by following the 'Create Ticket' link on that page.
To report a Maxima bug, you must have a Sourceforge account.
A problem in the graphical user interface is probably a wxMaxima bug.
The wxMaxima bug database is available at
    https://github.com/wxMaxima-developers/wxmaxima/issues
Submit bug reports by following the 'New issue' link on that page.
Please check before submitting, if your bug was already reported.
Please include the following information with your bug report:
-------------------------------------------------------------
wxMaxima version: 20.04.0_MAC
using wxWidgets version: wxWidgets 3.0.4
Maxima version: 5.43.2
Maxima build date: 2020-04-27 14:08:26
Host type: x86_64-apple-darwin19.4.0
System type: Darwin 19.4.0 X86-64
Lisp implementation type: SBCL
Lisp implementation version: 2.0.4
wxMaxima's idea of the directory layout is:
User configuration dir: /Users/msoegtrop/.maxima/
Help dir: /Applications/MacPorts/wxMaxima.app/Contents/Resources
-------------------------------------------------------------
```

**Comment by gunterkoenigsmann** [comment]
I tried to separate the size cache for all objects from the information that a size has already been determined a few days ago - and believe that this isn't the final solution to all related problems. But on my computer I cannot reproduce the crash anymore and therefore believe that at least this issue might be resolved.

**Comment by KubaO** [comment]
I had some ideas about making such caching automatic, but that requires another round of dependency tracking helper code. I prototyped something similar in the hierarchical style branch. But the very next task I want to accomplish is to wrap up the speed improvements as those will make it seem like a new product almost.

**Comment by gunterkoenigsmann** [comment]
The size caches always felt like being rather manual.  And at the speedups I am really looking forward to...

**Comment by MSoegtropIMC** [comment]
@gunterkoenigsmann : thanks - I will try and see if it works.

**Comment by KubaO** [comment]
@MSoegtropIMC Please reconfirm that this still happens for you on the master branch or in the nightly build. Otherwise I'll close it soon.

**Comment by MSoegtropIMC** [comment]
Sorry, I didn't have time to check this as yet - busy with the release of another open source projects. I likely won't have time before one or 2 weeks from now.

**Comment by MSoegtropIMC** [comment]
@gunterkoenigsmann : I just tested the latest commit 27d511afa59e0cad1ad87a4ffdbe41645b287576.

It still has this issue.

Testing this I also once got this assert:
```
/opt/local/Library/Frameworks/wxWidgets.framework/Versions/wxWidgets/3.0/include/wx-3.0/wx/ctrlsub.h(147): assert "pos <= GetCount()" failed in InsertItems(): position out of range

Call stack:
[00] wxItemContainer::InsertItems(wxArrayStringsAdapter const&, unsigned int) 
[01] wxItemContainer::Insert(wxString const&, unsigned int) 
[02] History::AddToHistory(wxString const&)  
[03] wxMaximaFrame::AddToHistory(wxString const&) 
[04] wxMaxima::UpdateDeferredHistory()       
[05] wxMaxima::OnIdle(wxIdleEvent&)          
[06] wxEvtHandler::ProcessEventIfMatchesId(wxEventTableEntryBase const&, wxEvtHandler*, wxEvent&) 
[07] wxEvtHandler::SearchDynamicEventTable(wxEvent&) 
[08] wxEvtHandler::TryHereOnly(wxEvent&)     
[09] wxEvtHandler::DoTryChain(wxEvent&)      
[10] wxEvtHandler::ProcessEvent(wxEvent&)    
[11] wxEvtHandler::SafelyProcessEvent(wxEvent&) 
[12] wxWindowBase::SendIdleEvents(wxIdleEvent&) 
[13] wxAppBase::ProcessIdle()                
[14] wxApp::ProcessIdle()                    
[15] wxCFEventLoop::CommonModeObserverCallBack(__CFRunLoopObserver*, int) 
[16] __CFRUNLOOP_IS_CALLING_OUT_TO_AN_OBSERVER_CALLBACK_FUNCTION__ 
[17] __CFRunLoopDoObservers                  
[18] __CFRunLoopRun                          
[19] CFRunLoopRunSpecific                    
[20] RunCurrentEventLoopInMode               
Do you want to stop the program?
You can also choose [Cancel] to suppress further warnings.
```

But it is not reproducible.

Can you reproduce this issue on Windows/Linux?

**Comment by KubaO** [comment]
I have an idea what's up. History code needs my review badly :)

**Comment by KubaO** [comment]
1. History issue: I have reworked History and it has such problems fixed.

2. Paste while selected: it simply doesn't reproduce for me. I need to know whether you get a crash or an assertion, and if you get a crash then without a backtrace I can't do much. But you'd need to test the master branch, whatever's most recent.

**Comment by MSoegtropIMC** [comment]
@KubaO : it usually crashes without an assert, but I can do a debug build and run it in the debugger. With master from July 11th this was 100% reproducible. You are also on Mac?

**Comment by KubaO** [comment]
I’m trying it on Windows, but I have a mac right next to me so I’ll try there too.

**Comment by KubaO** [comment]
I've found a problem where pasting when more than one cell is selected crashes. Maybe this is what you're referring to?

The fix will be merged soon.

**Comment by KubaO** [comment]
This should be hopefully fixed by 03b00ce5d8. Please test.

**Comment by MSoegtropIMC** [comment]
@KubaO : this is indeed fixed in master of today, so I will close the issue.
Thanks for fixing!
